### PR TITLE
Fix #11278 QueryBuilder::setParameter and ArrayParameterType

### LINKED ISSUE #11278: Doctrine\ORM\QueryBuilder::setParameter and ArrayParameterType in DBAL 4.0.0
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 3.0.0

#### Summary

`Doctrine\ORM\QueryBuilder::setParameter()` needs an optional parameter for complex data types sometimes and in 3.x version, it expects `string|int|null` and can be linked via `Doctrine\DBAL\Connection` e.g. `Doctrine\DBAL\Connection::PARAM_STR_ARRAY`. These values used to be as int datatype and have been deprecated. Now these constants are linked to `Doctrine\DBAL\ArrayParameterType` class. In 4.0.0 class ArrayParameterType has been converted to Enums but setParameter() is still accepting old `string|int|null` resulting following error

More info at https://github.com/doctrine/dbal/issues/6309

### PR DESCRIPTION
**Pull request by hanishsingla**

### CODE REVIEW

**Comment by greg0ire** [comment]
Should the same thing be done for `ParameterType` as well? It has also become an enum.

File: src/QueryBuilder.php [hunk 1]
```diff
@@ -428,12 +429,12 @@ public function getRootEntities(): array
      *         ->setParameter('user_id', 1);
      * </code>
      *
-     * @param string|int      $key  The parameter position or name.
-     * @param string|int|null $type ParameterType::* or \Doctrine\DBAL\Types\Type::* constant
+     * @param string|int                         $key  The parameter position or name.
+     * @param ArrayParameterType|string|int|null $type ParameterType::* or \Doctrine\DBAL\Types\Type::* constant or \Doctrine\DBAL\ArrayParameterType enum
```
Comment (derrabus) [inline src/QueryBuilder.php]:
```suggestion
     * @param ParameterType|ArrayParameterType|string|int|null $type ParameterType::*, ArrayParameterType::* or \Doctrine\DBAL\Types\Type::* constant
```

File: src/QueryBuilder.php [hunk 2]
```diff
@@ -428,12 +429,12 @@ public function getRootEntities(): array
...
      *
      * @return $this
      */
-    public function setParameter(string|int $key, mixed $value, string|int|null $type = null): static
+    public function setParameter(string|int $key, mixed $value, ArrayParameterType|string|int|null $type = null): static
```
Comment (derrabus) [inline src/QueryBuilder.php]:
```suggestion
    public function setParameter(string|int $key, mixed $value, ParameterType|ArrayParameterType|string|int|null $type = null): static
```

**Review by derrabus — CHANGES_REQUESTED** [review]
> Should the same thing be done for `ParameterType` as well? It has also become an enum.

Yes.

And maybe we should add tests that would've discovered this problem. 🙈

**Comment by hanishsingla** [comment]
@derrabus This is my first PR to doctrine, so need some guidance. Should I create a new PR with latest changes as per your suggestions or there are other steps involved? Thanks

**Comment by derrabus** [comment]
No need fir a new PR, just amend this one.

**Comment by derrabus** [comment]
Closing in favor of #11287.

**Pull request closed without merging**